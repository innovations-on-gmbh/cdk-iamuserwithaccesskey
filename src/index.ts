import { Fn, CfnOutput } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as sm from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';

/**
 * Properties for the IAM User
 */
export interface IamUserWithAccessKeyProps extends iam.UserProps {
  /**
   * An optional custom encryption key for the secret.
   *
   * @default The Accounts default Secret Manager KMS Key will be used.
   */
  readonly encryptionKey?: kms.IKey | undefined;
}

/**
 * An IAM User including an Access Key that will be stored in Secrets Manager. The properties as for normal IAM Users.
 */
export class IamUserWithAccessKey extends Construct {
  /**
   * An attribute that represents the iam access_key.
   *
   * @attribute true
   */
  public readonly accessKey: iam.CfnAccessKey;
  /**
     * An attribute that represents the secret.
     *
     * @attribute true
     */
  public readonly secret: sm.Secret;

  constructor(scope: Construct, id: string, props?: IamUserWithAccessKeyProps) {
    super(scope, id);

    const user = new iam.User(this, 'Resource', props);

    this.accessKey = new iam.CfnAccessKey(this, 'AccessKey', {
      userName: user.userName,
    });

    let UserSecretString = JSON.stringify({
      Access_Key_Id: this.accessKey.ref,
      Secret_Access_Key: Fn.getAtt(this.accessKey.logicalId, 'SecretAccessKey').toString(), //TODO: Check if this references the correct resource. AccessKey vs SecretAccessKey
    });

    if (props?.encryptionKey) {
      this.secret = new sm.Secret(this, 'UserSecret', {
        encryptionKey: props.encryptionKey,
      });
    } else {
      this.secret = new sm.Secret(this, 'UserSecret');
    };

    // We need to access the underlying cfn resource to set the secret string
    const cfnSecret = this.secret.node.defaultChild as sm.CfnSecret;
    cfnSecret.secretString = UserSecretString;

    // We need a raw override because otherwise cdk always expects a secretStringGenerator object
    cfnSecret.addOverride('Properties.GenerateSecretString', Fn.ref('AWS::NoValue'));

    new CfnOutput(this, 'SecretArn', {
      value: this.secret.secretArn.toString(),
      exportName: `${id}`,
    });
  }
}
