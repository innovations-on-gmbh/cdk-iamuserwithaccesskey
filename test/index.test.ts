import { expect, countResources } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { IamUserWithAccessKey } from '../src';

describe('User', () => {
  it('creates the user', () => {
    const stack = new Stack();
    new IamUserWithAccessKey(stack, 'User', {
      // secretName: 'IamUserWithAccessKeySecret',
      userName: 'iamUserWithAccessKeyUsername',
    });
    expect(stack).to(countResources('AWS::IAM::User', 1));
    expect(stack).to(countResources('AWS::IAM::AccessKey', 1));
    expect(stack).to(countResources('AWS::SecretsManager::Secret', 1));
    //TODO: Test that generateSecretString for the Secret is always null!
  });
});
