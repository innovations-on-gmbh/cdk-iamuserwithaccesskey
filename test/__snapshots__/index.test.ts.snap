// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Template has the correct resources 1`] = `
Object {
  "Outputs": Object {
    "UserSecretArn35117ED6": Object {
      "Export": Object {
        "Name": "User",
      },
      "Value": Object {
        "Ref": "UserUserSecret4251AFB1",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "UserAccessKeyCB7DC9A9": Object {
      "Properties": Object {
        "UserName": Object {
          "Ref": "UserFDDCDD17",
        },
      },
      "Type": "AWS::IAM::AccessKey",
    },
    "UserFDDCDD17": Object {
      "Properties": Object {
        "UserName": "iamUserWithAccessKeyUsername",
      },
      "Type": "AWS::IAM::User",
    },
    "UserUserSecret4251AFB1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "GenerateSecretString": Object {
          "Ref": "AWS::NoValue",
        },
        "SecretString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"Access_Key_Id\\":\\"",
              Object {
                "Ref": "UserAccessKeyCB7DC9A9",
              },
              "\\",\\"Secret_Access_Key\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "UserAccessKeyCB7DC9A9",
                  "SecretAccessKey",
                ],
              },
              "\\"}",
            ],
          ],
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
